### DISCUSSION TITLE
Alternative to `addPostFrameCallback` in `RearchConsumer`?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#49] [Q&A] [answered]
Is there any way to execute it on `initState` (callonce)?

**Comment by busslina** [comment]
I'm just thinking a fix:

```dart
extension _UseConvenience on WidgetSideEffectRegistrar {
  WidgetSideEffectRegistrar get use => this;
}

extension CustomSideEffects on WidgetSideEffectRegistrar {
  void afterBuild(void Function() callback) {
    final (buildCount, setBuildCount) = use.state(0);

    if (buildCount == 1) {
      callback();
    }

    setBuildCount(buildCount + 1);
  }
}
```

**Comment by GregoryConrad** [comment, marked as answer]
Using addPostFrameCallback is often a workaround for a different issue. 99% of the time there is a better alternative. Such as: use the rawValueWrapper instead of state to not trigger a rebuild.

that being said, you can put an addPostFrameCallback in a callonce() for the really rare times you need it (at least for an initState alternative)

**Reply by busslina** [reply]
Cool, I'm just trying to emulate a `State` extension method that I use a lot (`State.afterBuild`).

Corrected:
```dart
void afterBuild(void Function() callback) {
    final (getBuildCount, setBuildCount) = use.rawValueWrapper(() => 0);

    final buildCount = getBuildCount();

    if (buildCount == 1) {
      callback();
    }

    setBuildCount(buildCount + 1);
  }
```

**Reply by busslina** [reply]
> that being said, you can put an addPostFrameCallback in a callonce() for the really rare times you need it (at least for an initState alternative)

I assume you mean to use my custom `addPostFrameCallback` because it's not available inside `RearchConsumer`

**Reply by GregoryConrad** [reply]
No, I meant the addPostFrameCallback available on the WidgetScheduler (or whatever it is called, can’t remember). Alternatively you can do an await null; or Future.delayed to schedule a new future that will execute after the current frame.

your afterBuild method will only execute on the second build and a rebuild would actually need to be triggered elsewhere for it to execute.

**Reply by busslina** [reply]
Yeah, it's on purpose to only execute on the second build (I didn't choose the best name for it). Basically I was calling `addPostFrameCallback` from `initState` once.
